scRNA-seq/02-normalizing_scRNA-seq.Rmd

```diff
@@ -67,185 +72,355 @@ sc_metadata <- readr::read_tsv(metadata_file)
+
+```
+
+Because we tend to like [tidy data](https://r4ds.had.co.nz/tidy-data.html), here we use a `tidyverse` function from the [`purrr` package](https://purrr.tidyverse.org) to apply the same operations as above to every element of the `markers` list. 
+We will introduce `purrr` briefly here, but if you want more information and background, we recommend the [`purrr` cheatsheet (PDF)](https://github.com/rstudio/cheatsheets/raw/master/purrr.pdf) and Jenny Bryan's great [`purrr` tutorial](https://jennybc.github.io/purrr-tutorial/index.html) you might want to look at.
+
+
+The main functions in `purrr` are the `map()` functions, which take as their main arguments a **list** and a **function** to apply to each element of the list. 
+The first one, `purrr::map()` is essentially equivalent to the `lapply()` function that is part of base R, but with some different defaults. 
+We will use it to get the top rows from each table by applying the `head()` function to each element of the list.
+The results are returned as a new list.
+
+```{r head_markers, eval = FALSE}
+purrr::map(as.list(markers), # convert markers to a 'regular' list for purrr
+           head) # the function to apply (note no parenthesis)
+```
+
+This returns a list of data frames, which isn't quite what we want.
+
+There is no built-in function that will give us just the first few row names, so we will have to define one, and `purrr` gives us a nice shorthand for doing that with `~` syntax.
+In this syntax we type a `~` followed by an expression (R code) that uses `.x` as a placeholder for a single element of the list.
+`purrr::map()` will then apply the expression to each element of the list, and return the results as a new list.
+
+```{r head_markernames, live = TRUE}
+# Get the first few row names of each table with a purrr function.
+purrr::map(as.list(markers), # convert markers to a 'regular' list for purrr
+           ~ rownames(head(.x)) )#  our custom function.
+```
+
+Another variant is `purrr::imap()`, which allows us to use the names of the list elements in our function. 
+(UTry`names(markers)` to see the names for the list we are working with now.)
+We will use that here to name output files where we will print each of the marker tables, one for each cell type.
+We are again defining a custom function within the call to `purrr:imap()` using the `~` syntax, but this time we can use `.x` for the list elements and `.y` for their names.
+
+In addition to writing the tables out, we are saving the data frames we created as a new list object that we can use in the next step.
+
+```{r write_tables}
+marker_df_list <- purrr::imap(
+  as.list(markers), # convert markers to a 'regular' list for purrr
+  ~ as.data.frame(.x) %>% # convert the table to a data frame
+    tibble::rownames_to_column("gene") %>% # make genes a new column
+    readr::write_tsv( # write each data frame to a file
+      file.path(marker_dir, paste0(.y, "_markers.tsv")) # construct the output path
+      )
```

Comment:
```suggestion
     )
```
I think?

Reply:
this one looks right to me too?

Reply:
Never mind, you are correct, I was a tab too far. 📏